src/library/scala/concurrent/impl/ExecutionContextImpl.scala

```diff
@@ -53,39 +52,42 @@ private[concurrent] object ExecutionContextImpl {
       thread
     }
 
-    // As per ThreadFactory contract newThread should return `null` if cannot create new thread.
-    def newThread(runnable: Runnable): Thread =
-      if (reserveThread())
-        wire(new Thread(new Runnable {
-          // We have to decrement the current thread count when the thread exits
-          override def run() = try runnable.run() finally deregisterThread()
-        })) else null
+    def newThread(runnable: Runnable): Thread = wire(new Thread(runnable))
 
     def newThread(fjp: ForkJoinPool): ForkJoinWorkerThread =
-      if (reserveThread()) {
-        wire(new ForkJoinWorkerThread(fjp) with BlockContext {
-          // We have to decrement the current thread count when the thread exits
-          final override def onTermination(exception: Throwable): Unit = deregisterThread()
-          final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = {
-            var result: T = null.asInstanceOf[T]
-            ForkJoinPool.managedBlock(new ForkJoinPool.ManagedBlocker {
-              @volatile var isdone = false
-              override def block(): Boolean = {
-                result = try {
-                    // When we block, switch out the BlockContext temporarily so that nested blocking does not created N new Threads
-                    BlockContext.withBlockContext(BlockContext.defaultBlockContext) { thunk }
-                  } finally {
-                    isdone = true
+      wire(new ForkJoinWorkerThread(fjp) with BlockContext {
+        private[this] var isBlocked: Boolean = false // This is only ever read & written if this thread is the current thread
+        final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T =
+          if ((Thread.currentThread eq this) && !isBlocked && newBlocker()) {
+            try {
+              isBlocked = true
+              val b: ForkJoinPool.ManagedBlocker with (() => T) =
+                new ForkJoinPool.ManagedBlocker with (() => T) {
+                  private[this] var result: T = null.asInstanceOf[T]
+                  private[this] var done: Boolean = false
+                  final override def block(): Boolean = {
+                    try {
+                      if (!done)
+                        result = thunk
+                    } finally {
+                      done = true
+                    }
+
+                    true
                   }
 
-                true
-              }
-              override def isReleasable = isdone
-            })
-            result
-          }
-        })
-      } else null
+                  final override def isReleasable = done
+
+                  final override def apply(): T = result
+                }
+              ForkJoinPool.managedBlock(b)
+              b()
```

Comment:
I'm curious as to the rationale behind why this is defined as an `apply()` method, instead of moving `result` outside of `b`

Reply:
To avoid having the compiler having to allocate an additional holder for the value.